storage/badger/store.go

```diff
@@ -21,43 +21,236 @@ package badger
-	"github.com/onflow/flow-go/engine/execution/state/delta"
-	flowgo "github.com/onflow/flow-go/model/flow"
-
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/google/uuid"
 	"github.com/onflow/flow-emulator/storage"
 	"github.com/onflow/flow-emulator/types"
+	"github.com/onflow/flow-go/engine/execution/state/delta"
+	flowgo "github.com/onflow/flow-go/model/flow"
 )
 
 // Store is an embedded storage implementation using Badger as the underlying
 // persistent key-value store.
 type Store struct {
 	db              *badger.DB
 	ledgerChangeLog changelog
+	dbGitRepository *git.Repository
+	path            string
+	badgerOptions   badger.Options
 }
 
 var _ storage.Store = &Store{}
 
+func getTag(r *git.Repository, tag string) *object.Tag {
+	tags, err := r.TagObjects()
+	if err != nil {
+		return nil
+	}
+	var res *object.Tag = nil
+	tags.ForEach(func(t *object.Tag) error {
+		if t.Name == tag {
+			res = t
+		}
+		return nil
+	})
+	return res
+}
+
+func setTag(r *git.Repository, tag string, tagger *object.Signature) (bool, error) {
+	if getTag(r, tag) != nil {
+		return false, nil
+	}
+	h, err := r.Head()
+	if err != nil {
+		return false, err
+	}
+	_, err = r.CreateTag(tag, h.Hash(), &git.CreateTagOptions{
+		Tagger:  tagger,
+		Message: tag,
+	})
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+func defaultSignature(name, email string) *object.Signature {
+	return &object.Signature{
+		Name:  name,
+		Email: email,
+		When:  time.Now(),
+	}
+}
+
+//prevents git commits when emulator running
+func (s *Store) lockGit() {
+	lockPath := fmt.Sprintf("%s/.git/index.lock", s.path)
+	ioutil.WriteFile(lockPath, []byte("emulatorLock"), 0755)
+}
+
+func (s *Store) unlockGit() {
+	lockPath := fmt.Sprintf("%s/.git/index.lock", s.path)
+	os.Remove(lockPath)
+}
+
+func (s *Store) JumpToContext(context string) error {
+	s.unlockGit()
+	defer s.lockGit()
+	err := s.db.Close()
+	if err != nil {
+		return err
+	}
+
+	s.newCommit(fmt.Sprintf("Context switching to: %s", context))
+
+	w, _ := s.dbGitRepository.Worktree()
+
+	branch := fmt.Sprintf("refs/heads/%s", context)
+	b := plumbing.ReferenceName(branch)
+
+	// checkout branch ( first branch name is actualy context name )
+	err = w.Checkout(&git.CheckoutOptions{Create: false, Force: true, Branch: b})
+
+	if err != nil {
+		// branch doesn't exist, it means we need to create it ( first branch is named after context )
+		err := w.Checkout(&git.CheckoutOptions{Create: true, Force: true, Branch: b})
+		if err != nil {
+			return err
+		}
+
+		//after we create a tag pointing to start of this context
+		created, err := setTag(s.dbGitRepository, context, defaultSignature("Emulator", "emulator@onflow.org"))
+		if err != nil && !created {
+			return err
+		}
+
+	} else {
+
+		//create new branch
+		uuidWithHyphen := uuid.New()
+		uuid := strings.Replace(uuidWithHyphen.String(), "-", "", -1)
```

Comment:
conflicts with imported package

Reply:
can you explain this a bit more?

Reply:
just the fact that uuid is name of imported package, and it's better idea to name a variable differently to not conflict the namespace.

Reply:
ah ok thanks